### PR TITLE
persist: Handle the special case of columnar encoding an empty `RelationDesc`

### PR DESCRIPTION
**Pull request by ParkMyCar**
We encode `Row`s in Arrow with a `StructArray`, but Arrow doesn't support a `StructArray` with no fields. To handle this case [we use a `NullArray`](https://github.com/MaterializeInc/materialize/blob/main/src/persist-types/src/dyn_struct.rs#L97) on encode, but on decode we don't consider this case and [always expect a `StructArray`](https://github.com/MaterializeInc/materialize/blob/main/src/persist-types/src/dyn_struct.rs#L241).

This PR adds support for columnar encoding an empty `RelationDesc` in the new columnar encoder at the `SourceData` level. What makes this problem a little tricky is differentiating between an empty `Row` and a null entry. There are two approaches we could take to solve this problem:

1. Handle this in `RowColumnarEncoder`. We could add a special "internal" column when we create a `RowColumnarEncoder` for an empty `RelationDesc`, e.g. a `BooleanArray`. We could encode empty `Row`s as `true` and null entries as normal by setting an entry in the validity bitmap.
2. Handle this in `SourceDataColumnarEncoder`. When creating a `SourceDataColumnarEncoder` for an empty `RelationDesc`, instead of creating a `RowColumnarEncoder` we create a `NullArray`. On decode if the `err` column is null then we know to decode an empty `Row`.

This PR implements approach 2. A `NullArray` can be encoded in constant space, just an integer count for the number of nulls, and IMO is a bit cleaner than a special "internal" column.

I also only implemented this fix in the new columnar encoder since the existing one will be removed soon. I also added a proptest that generates an arbitrary `RelationDesc` and then generates arbitrary data for that `RelationDesc` and asserts it all roundtrips through Parquet. I ran this test for >48 hours.

### Motivation

Forward fixes https://github.com/MaterializeInc/materialize/issues/27840

### Tips for reviewer

This PR is split into two commits which can be reviewed separately:

1. The changes to support encoding an empty `RelationDesc`
4. Adds new proptest strategies and tests to add coverage for this case and more.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: src/repr/src/row/encoding2.rs [hunk 1]
```diff
@@ -615,10 +615,10 @@ impl DatumColumnEncoder {
...
                 let nulls = nulls.as_mut().map(|n| NullBuffer::from(n.finish()));
 
-                let key_field =
-                    Arc::new(Field::new("key", keys.data_type().clone(), nulls.is_some()));
-                let val_field =
-                    Arc::new(Field::new("val", vals.data_type().clone(), nulls.is_some()));
+                // The keys and values in a Map can never be null, instead a
+                // field being null is maintained by the higher level MapArray.
+                let key_field = Arc::new(Field::new("key", keys.data_type().clone(), false));
+                let val_field = Arc::new(Field::new("val", vals.data_type().clone(), false));
```
Comment (ParkMyCar) [inline src/repr/src/row/encoding2.rs:621]: this is unrelated to the issue of empty `RelationDesc`s, but was found with the new proptests.

File: src/repr/src/relation.rs [hunk 2]
```diff
@@ -562,6 +565,48 @@ impl RelationDesc {
     }
 }
 
+impl Arbitrary for RelationDesc {
```
Comment (bkirwi) [inline src/repr/src/relation.rs:568]: 🆒

**Review by bkirwi — COMMENTED** [review]
> There are two approaches we could take to solve this problem:

To me the compelling bit is that we try and use the same representation for empty rows as we do for empty records. My understanding is that we use `BooleanArray` in that case? `NullArray` like you have here seems more appropriate to me, though. Possible (and not too onerous) to bring those two things in line?

**Comment by ParkMyCar** [comment]
> > There are two approaches we could take to solve this problem:
> 
> To me the compelling bit is that we try and use the same representation for empty rows as we do for empty records. My understanding is that we use `BooleanArray` in that case? `NullArray` like you have here seems more appropriate to me, though. Possible (and not too onerous) to bring those two things in line?

I agree that `NullArray` is the better approach unfortunately, it doesn't work for empty Record because there is no way to distinguish a valid entry with an empty Record vs a null entry. Specifically when decoding `SourceData` where the `RelationDesc` is empty, we rely on the absence of a value in the 'err' column to indicate that  we should decode an empty Row. But for a column of empty Records we don't have similar metadata that we can rely on which is why we need to use the `BooleanArray` approach.

**Comment by bkirwi** [comment]
Ugh, I had believed that `NullArray` had validity information attached, but of course it does not! (And I think you mentioned in the past...)

I think my feeling is still that uniformity is worth it, but it's no longer strong enough that I'd suggest rewriting anything. 😅 Thanks for talking through it!

**Review by bkirwi — APPROVED** [review]

**Comment by ParkMyCar** [comment]
> Ugh, I had believed that `NullArray` had validity information attached, but of course it does not! (And I think you mentioned in the past...)

Given the `...Array` naming you would think it does! Unfortunately it's more or less just a count of the number of nulls.

> I think my feeling is still that uniformity is worth it, but it's no longer strong enough that I'd suggest rewriting anything. 😅 Thanks for talking through it!

Of course, thanks for the thorough review!

**Pull request merged**